workers/unity/Assets/Gdk/Core/Components/Option.cs

```diff
@@ -0,0 +1,116 @@
+using System;
+
+namespace Improbable.Gdk.Core
+{
+    public struct Option<T> : IEquatable<Option<T>>
+    {
+        public static readonly Option<T> Empty = new Option<T>();
+
+        public BlittableBool HasValue { get; }
+
+        private readonly T value;
+
+        public T Value
+        {
+            get
+            {
+                if (!HasValue)
+                {
+                    throw new CalledValueOnEmptyOptionException("Called Value on empty Option.");
+                }
+
+                return value;
+            }
+        }
+
+        public Option(T value)
+        {
+            HasValue = true;
+            this.value = value;
+        }
+
+        public bool TryGetValue(out T value)
+        {
+            if (!HasValue)
+            {
+                value = default(T);
+                return false;
+            }
+
+            value = this.value;
+            return true;
+        }
+
+        public T GetValueOrDefault(T defaultValue)
+        {
+            return HasValue ? value : defaultValue;
+        }
+
+        public override bool Equals(object other)
+        {
+            return other is Option<T> && this.Equals((Option<T>) other);
+        }
+
+        public bool Equals(Option<T> other)
+        {
+            if (!HasValue && !other.HasValue)
+            {
+                return true;
+            }
+
+            if (HasValue && other.HasValue)
+            {
+                return value.Equals(other.Value);
```

Comment:
Does this mean we can only make options of equatable structs?

Reply:
Actually looked it up. The answer is no, but this is a boxing allocation. Do we use equality in the updates structs anywhere?

Reply:
Just tested this. This should work for everything except for options with null payloads where you will attempt to call `null.Equals(something else)`. That I can fix. Everything else, value types, reference types, structs, should provide a default Object.Equals method.

Reply:
Oh shit. It's the boxing problem all over again.